compiler/src/formatting/format.re

```diff
@@ -744,238 +760,251 @@ let rec block_item_iterator =
+            [
+              Doc.hardLine,
+              Comment_utils.nobreak_comment_to_doc(comment),
+              ...acc,
+            ];
+          } else {
+            force_break_for_comment(comment, acc);
+          };
 
-          let filler =
-            switch (last_comment) {
-            | Line(_) => Doc.nil
-            | _ => Doc.space
-            };
-          Doc.concat([
-            Comment_utils.new_comments_to_docs(leading_comments),
-            filler,
-            print_item(~comments=item_comments, item),
-            separator,
-          ]);
+        | Some(next) =>
+          let (_, next_comment_line, _, _) =
+            Locations.get_raw_pos_info(
+              Locations.get_comment_loc(next).loc_end,
+            );
 
-        | ([], [trailing, ..._]) =>
-          Doc.concat([
-            print_item(~comments=item_comments, item),
-            separator,
-            bcb,
-            Comment_utils.block_trailing_comments_docs(trailing_comments),
-            after_comments_break,
-          ])
+          next_comment := Some(comment);
 
-        | ([], []) =>
-          Doc.concat([print_item(~comments=item_comments, item), separator])
+          if (this_comment_line < next_comment_line) {
+            [
+              Doc.hardLine,
+              Comment_utils.nobreak_comment_to_doc(comment),
+              ...acc,
+            ];
+          } else {
+            force_break_for_comment(comment, acc);
+          };
         };
+      },
+      comments,
+      [],
+    );
 
-      [
-        this_item,
-        ...item_iterator(
-             ~previous=get_loc(item),
-             ~get_loc,
-             ~print_item,
-             ~comments,
-             ~separator,
-             ~followed_by_arrow?,
-             remainder,
-           ),
-      ];
+  if (separated^) {
+    Doc.concat(items);
+  } else {
+    Doc.ifBreaks(
+      Doc.concat([separator, Doc.concat(items)]),
+      Doc.concat([Doc.concat(items), separator]),
+    );
+  };
+};
 
-    | [] =>
-      let comments_without_item_comments =
-        remove_used_comments(~remove_comments=item_comments, comments);
-      let trailing_comments =
-        Comment_utils.get_comments_after_location(
-          ~location=get_loc(item),
-          comments,
-        );
+let rec item_iterator =
+        (
+          ~previous: option(Location.t)=?,
+          ~get_loc: 'a => Location.t,
+          ~print_item: (~comments: list(Parsetree.comment), 'a) => Doc.t,
+          ~comments: list(Parsetree.comment),
+          ~followed_by_arrow: option(bool)=?,
+          ~iterated_item: iterator_item_type,
+          items: list('a),
+        ) => {
+  let trailing_separator =
+    switch (iterated_item) {
+    | IteratedListPattern
+    | IteratedDataDeclarations
+    | IteratedMatchItem
+    | IteratedRecordLabels => true
+    | IteratedRecordPattern
+    | IteratedRecord
+    | IteratedTypeConstructor
+    | IteratedPatterns
```

Comment:
when is this?

Reply:
Sorry, can you give more detail?

Reply:
Just asking what type of code that references!